### ISSUE TITLE
Sensor relocation following sorghum 4 harvest

### ISSUE THREAD

**Issue by smarshall-bmr** [terraref/computing-pipeline#352]
This issue is to serve as a warning and discussion area for upcoming changes to sensor locations in the sensor box. I'm not sure of all the methods the downstream extractors are using but be prepared for slightly differently structured data after this season!

Sensors "confirmed" to be moved are:

- Top 3D laser scanners will be moved around 2cm south in the sensor box
- NDVI/PRI and CROPCircle will be moved so they share y-axis coordinates with the top StereoVis
- Side 3D lasers moved... somewhere not where they are now

Sensors "in priority discussion" for possible movement are:

- FLIR to better clear sensor box shade
- VNIR/SWIR for shading issues

Other sensors being discussed:

- possibly moving/repurposing side 3D laser to sensor box for x-axis scanning
- relocating side StereoVis to support photogrammetry/ work better with x-axis scan paradigm
- any other suggestions?

Is there a sensor change or addition that would make your life easier or would improve the output?
What is working well now/ could possibly break extractors that we shouldn't change?

Ideally I'd like to hear from everyone downstream (even if you think the gantry is perfect the way it is and you don't want anything more! 👍 )

- [ ] move top 3D

- [ ] move PRI/NDVI cluster

- [ ] move CROPCircle

- [ ] decide on FLIR and move if necessary

- [ ] decide on SWIR/VNIR and move if necessary

- [ ] discuss/decide on other sensors

**Comment by dlebauer** [comment]
Would it be possible to install the GoPros at the same time?

I would suggest the following timeline:

* ASAP: Decide which sensors to move
* At least 1 month before next planting season:
    * Move sensors
    * Update  fixed metadata
    * Run test scans
    * evaluate output, especially for geolocation

**Comment by rachelshekar** [comment]
@ZongyangLi and @solmazhajmohammadi please comment

**Comment by mabitter-x** [comment]
For many machine learning applications it would be useful to have a single image taken of each plot in its entirety.  This will be true of all of the snapshot-capture imagers.  This would encompass the StereoTop, Thermal, PRI/NDVI multispec, and even the pushbroom style hyspec imagers depending on the way you have it set up.    

The value of this cannot be understated.  Images of plants in the wind is a very difficult thing to stitch together especially as the canopy closes, and despite how well it can theoretically be done there are other unavoidable relics that will throw off remote sensing calculations - things like leaves and grain appearing in multiple images within the same plot due to sway from wind, and the relics of any image manipulation throwing off downstream computational processes.  

This would be more valuable to us at this point in our research than the hockney of detailed shots, as we think it help achieve higher accuracy ground-truthing of images.  We'd love to see it in addition to the detailed close-up work, however would take it in lieu if we had to choose. 

Thanks for all your awesome hard work, and your continual dedication to improving this fantastic tool and dataset.  It really is the hubble telescope for agriculture, and we're excited to see all that is learned from it.  

Best!  - Matthew

**Comment by dlebauer** [comment]
* PRI, NDVI x2, PAR sensor, spectrometer, and CO2 sensor have all been removed and are either in Stuart’s custody or shipped to manufacturer for factory recalibration.
* FLIR will be recalibrated in Arizona, under supervision of Andy French. Development of protocols will be done by the TERRA-REF team ( @remotesensinglab and @solmazhajmohammadi ) , Andy will provide test chamber and software

Once back from repairs, the recalibration will be performed by Stuart, with supervision from Andy

**Comment by NewcombMaria** [comment]
The VNIR is currently located on the West side of the camera box and Stuart mentioned that the VNIR may not be able to fully image the most Eastern experimental plots with the field planting arrangement that we've been following.  If the VNIR is relocated, can it be shifted eastward so that it can image all of the experimental plots across the Y axis?  (another alternative would be to extend the Y axis motion limit setting if that's a possibility after the side-arm sensors are removed).

**Comment by NewcombMaria** [comment]
@smarshall-bmr and @dlebauer we met as a local group earlier this week and discussed a written pre-season protocol/checklist for scanner system and field optimizations before data collection.  We came up with the following 6 steps that need to be completed before relocating any sensors.  Please review and suggest any edits/improvements/additions.  

Between-season relocation of sensors/cameras, accounting for the full set of downstream adjustments and tests that are required: 
	Step 1 – pre-relocation test scan of correct target with sensor in current position
	Step 2 – relocate, with emphasis on precise leveling (maintaining horizontal position)
	Step 3 – post-relocation test scan of correct target with sensor in new position
	Step 4 – image/data analysis in coordination with Illinois computing team and/or Solmaz as needed to compare expected shift to realized shift and ensure ability to fuse data among cameras)
	Step 5 – change offsets and relocation notes in metadata
	Step 6 – change scripts, focal distances if needed, sun/shade settings if needed
	Step 7 – test scripts with new adjustments

**Comment by max-zilla** [comment]
need to keep info in fixed metadata for previous and new time ranges.

**Comment by TinoDornbusch** [comment]
As already stated, within the year1 support contract, I could help the Mac team to establish some local scripts (based on Matlab/Octave) for simple feature extraction for quality control  onsite for the different sensors.

Visualisation of data could either be done in Matlab or with Microsoft tools (Power BI). This would be relatively easy and straightfoward.

This would allow faster response time, for sensor testing onsite.

**Comment by JeffWhiteAZ** [comment]
Would there be a need for new scripts if the proposed Python scripts for each sensor are sufficient? It might help our QC efforts at MAC if we were using the same tools as used for the main processing work flow.

-          Jeff

**Comment by dlebauer** [comment]
@TinoDornbusch it would be useful to have QA/QC scripts that we can use in the overall pipeline, for which Octave would work. Within the scope of this issue, it would be useful if you could write scripts related to analyzing the targets to confirm that we are seeing what we expect to see. However, it would be better to evaluate the level 1 products generated by the pipeline at NCSA as those have accounted for the sensor metadata, environmental conditions, changes in position, and etc.

@JeffWhiteAZ although Python orchestrates the main pipeline, it is by design sufficiently flexible to use a wide range of software and algorithms .... Octave (or R, Python, C/C++, etc) would be easiest to integrate and most broadly useful to the community since they are open source and run on Linux. Matlab would work but would add the overhead of requiring a license (which we have at UI). Microsoft Power BI could be worth it if the tools already exist. But agree that there it seems difficult to justify writing scripts that duplicate the pipeline and algorithms that we already have.

**Comment by NewcombMaria** [comment]
Group discussion during today's Tuesday teleconference resulted in decision to relocate the VNIR camera at this time.  For all other sensors recommended to be moved, this time can be used to come up with detailed plans to be prepared to do the relocation immediately after the wheat harvest.  During the VNIR relocation, is it possible to shift it a bit to the east to be able to scan the entire plot on the eastern most experimental column?

**Comment by dlebauer** [comment]
@rjstrand could you please upload the new proposed positions

Will start w/ moving VNIR/SWIR south

**Comment by remotesensinglab** [comment]
Is fusion of VNIR-SWIR with stereo important? Who are working on it? Original layout of sensors was designed to make sure having the same FOV for both Stereo RGB and VNIR/SWIR. Do we still need that?

1) VNIR+SWIR - align in the center, and move south,  reduces shadow, increases sunlight coverage.

2) If possible, we can move Stereo RGB and FLIR as well, to south?

3) If we can not move Stereo RGB to south, it should not be a problem because fusion can be done afterwards on higher level products (e.g., Level 2)

Target: 

(1) having an spectralon target extending outside the camera box. Plant height and operational hazards may be the issues. It can be designed as flexible and extensible to overcome the plant height issue?

(2) An extended target from the gantry arm. 

(3) Low maintenance ceramic tiles on the ground.

**Comment by JeffWhiteAZ** [comment]
On having RGB aligned w hyperspectral cameras. Presumably the original intention of having the same FOV was:
1. To facilitate comparisons of RGB-based phenotypes (cover, red-green index ...) with vegetation indices such as NDVI from the VNIR.
2. To allow creating a 3-D reconstruction that could assist in analyses of any hyperspectral data (e.g., correcting for leaf angle)
Intention number 1 seems like a low priority.
Has anyone attempted to use the RGB-stereo images to estimate leaf angle? Presumably, we need this for the FLIR as well, so if we move the FLIR, there is an argument for moving the RGB with the FLIR.

**Comment by dlebauer** [comment]
proposed locations from @rjstrand 

![camera arrangment 001](https://user-images.githubusercontent.com/464871/37356240-7e448d14-26b3-11e8-8362-76c5f9b15472.jpeg)

![camera arrangment 002](https://user-images.githubusercontent.com/464871/37356239-7e30fc0e-26b3-11e8-854c-997a58bc42fc.jpeg)

![camera arrangment 003](https://user-images.githubusercontent.com/464871/37356238-7e25085e-26b3-11e8-91ba-16da447be73e.jpeg)

![camera arrangment 004](https://user-images.githubusercontent.com/464871/37356235-7e1706b4-26b3-11e8-9b83-70fff504c22e.jpeg)

**Comment by JeffWhiteAZ** [comment]
On targets.
Design-wise, I think we can mount an arm with a small target, holding the arm in place using magnets. We can at least use such an assembly in wheat or other crops with low height variation and with sorghum before the crop gets too tall.
Have we heard from image analysis people about possible issues with having the arm+panel in the FOV of other images?
For ground-based targets, it would be good to have something with contrast that can be used for checking image resolution. Kelly Thorp has expressed an interest in having a summer intern look at different options with a focus on UAV targets.

**Comment by dlebauer** [comment]
Here are some illustrations of the 3 options summarized by @remotesensinglab in https://github.com/terraref/computing-pipeline/issues/352#issuecomment-372727152



1. spectralon target extending outside the camera box
2. extended target from the gantry arm.
3. Low maintenance ceramic tiles on the ground.

(I am interpreting 'on the ground' as 'staked into the ground')

<img width="245" alt="image" src="https://user-images.githubusercontent.com/464871/37357672-0cd38b22-26b7-11e8-9fc6-aa60f00c5c6f.png">

**Comment by JeffWhiteAZ** [comment]
Maria can comment with more knowledge of scan heights and reach, but I don't think option 2 is viable. The arm would have to be quite long (in Y = East-west direction) because the hyperspectral cameras only scan to within a certain distance. Unless it were above a persons height, it thus would be a substantial safety issue.
I favor a combination of the box-mounted arm and fixed-position targets.

**Comment by NewcombMaria** [comment]
Regarding targets, which are needed independently of sensor relocation, I've talked with people about the engineering logistics and it seems in theory that any of the 3 options that @dlebauer illustrated in the March 13 comment could work, with preference for the first option of extending an arm down from the camera box, less than 2 meters from the sensors to definitely clear the tops of plants.

**Comment by remotesensinglab** [comment]
If the option 1 is feasible, it would be a preferred solution.

**Comment by dlebauer** [comment]
For sensor relocation, today we discussed a few alternatives in this pptx file
[Camera arrangment without lighting.pptx](https://github.com/terraref/computing-pipeline/files/1853700/Camera.arrangment.without.lighting.pptx)

And agreed to move forward with configuration 4a

![image](https://user-images.githubusercontent.com/464871/37993409-3c9684ae-31d4-11e8-90cb-78a474b04253.png)

**Comment by NewcombMaria** [comment]
Yesterday Bob, John and I moved the camera box to the east limit and looked at alignment with the eastern-most experimental row.  We confirmed that the position of the FLIR in configuration 4a captures the entire experimental plot on the east side within the field of view of the FLIR.

**Comment by NewcombMaria** [comment]
Relocations completed 2018-08-18,  a modification of Configuration 4a. 
Photo taken from the ground looking up: 
![image](https://user-images.githubusercontent.com/22008637/44320669-56e70e00-a3f8-11e8-9f05-289d8e93070a.png)
